cli/lock.ml

```diff
@@ -209,20 +171,107 @@ let get_pin_depends ~global_state local_opam_files =
   root_pin_depends local_opam_files >>= pull_pin_depends ~global_state
 
 let calculate_opam ~build_only ~allow_jbuilder ~local_opam_files ~ocaml_version
-    =
+    ~target_packages =
   let open Rresult.R.Infix in
   OpamGlobalState.with_ `Lock_none (fun global_state ->
       OpamSwitchState.with_ `Lock_none global_state (fun switch_state ->
           check_repo_config ~switch_state;
           get_pin_depends ~global_state local_opam_files >>= fun pin_depends ->
           Opam_solve.calculate ~build_only ~allow_jbuilder ~local_opam_files
-            ~pin_depends ?ocaml_version switch_state))
+            ~target_packages ~pin_depends ?ocaml_version switch_state))
+
+let subselect_explicit ~local_packages ~explicit_list =
+  let by_opam_name entry (explicit_name, _) =
+    let name = entry |> Package_argument.name |> OpamPackage.Name.to_string in
+    String.equal name explicit_name
+  in
+
+  List.fold_left
+    ~f:(fun acc explicit ->
+      match (List.find_opt ~f:(by_opam_name explicit) local_packages, acc) with
+      | None, Ok _ -> Error [ explicit ]
+      | None, Error errors -> Error (explicit :: errors)
+      | Some _, Error errors -> Error errors
+      | Some x, Ok xs -> Ok (x :: xs))
+    ~init:(Ok []) explicit_list
+  |> Result.map_error ~f:(fun missing_packages ->
+         let msg =
+           Fmt.str "Packages %a specified but not found in repository"
+             Fmt.(list ~sep:comma Package_argument.pp_styled)
+             missing_packages
+         in
+         `Msg msg)
+
+let only_return_root xs =
+  let root = Sys.getcwd () |> Fpath.v in
+  List.filter
+    ~f:(fun (_, path) ->
+      match Fpath.relativize ~root path with
+      | None -> false
+      | Some path ->
+          let filename = path |> Fpath.filename |> Fpath.v in
+          Fpath.equal filename path)
+    xs
+
+let construct_version (name, repo) = (name, (None, repo))
+
+let group ~equal xs =
+  let rec loop acc xs =
+    match (xs, acc) with
+    | [], _ -> acc
+    | x :: xs, [] | x :: xs, [] :: _ -> loop [ [ x ] ] xs
+    | x :: xs, (y :: ys) :: zs -> (
+        match equal x y with
+        | true -> loop ((x :: y :: ys) :: zs) xs
+        | false -> loop ([ x ] :: (y :: ys) :: zs) xs)
+  in
+  loop [] xs
+
+let filter_unique xs =
+  let unique, dups = List.partition ~f:(fun x -> List.length x = 1) xs in
+  (match List.length dups with
+  | 0 -> ()
+  | _ ->
+      (* TODO: print duplicates in a less terrible way *)
+      let names = List.map ~f:(List.map ~f:fst) dups in
+      Fmt.epr "Duplicates removed: %a\n"
+        Fmt.(list ~sep:(any "; ") (list ~sep:(any " & ") string))
+        names);
+  unique
+
+let deduplicate_packages l =
+  let cmp (a, _) (b, _) = String.compare a b in
+  let equal (a, _) (b, _) = String.equal a b in
+  List.sort ~cmp l |> group ~equal |> filter_unique |> List.concat
+
+let target_path_map packages =
+  packages |> deduplicate_packages |> String.Map.of_list_exn
```

Comment:
It seems a bit weird to operate the deduplication on the list and to build a map after that. This can be done while building the map itself using something like `Map.update`.

Reply:
I thought about this but it is dependant on the semantics. If we want to remove all duplicated packages instead of just ignoring everything past the first one we encounter then this has to be done separately since otherwise we'll get into an odd edge case:

1. Package `a` found. We add it to the map
2. Another package `a` found. We attempt to add it to the map, realize it is already there so we remove both.
3. Another package `a` found. The map had its previous `a` removed, so we think this is the first time we see it, so we remove it.

This could be done with an additional set to track what we have already excluded but that feels to me quite roundabout, whereas this way we get all the duplicated packages in a data structure and can print them in one go so the user can get good feedback on what is wrong.

The code here is not great though, admittedly, with `deduplicate_packages` just doing some side-effect willy-nilly.

Reply:
I think a good behaviour is to keep at most one of each packages, meaning if there is a duplicate, we keep the first one we encountered, report about the others. Accumulating both a map and a list of ignored duplicates doesn't sound too bad to me.
It would have the advantage that we can report duplicates all at once in a more helpful manner, something like:
```
Package x is defined multiple times in the repo:
- ./path/to/first/x.opam
- ./path/to/second/x.opam
We kept ./path/to/first/x.opam and discarded the others
```

How does that sound?

We can still improve this in the future when we have an idea if this actually shows up and why!

I think in the end the proper fix will be to use some dune magic to know which packages are actually considered by dune at which point we can safely reject the repo as `dune build` won't work anyway.

I'm digressing a bit but my point is that there is no simple yet perfect behaviour here so let's start with the simplest one and we'll improve it overtime when we have a better idea about what the problem actually is.